### ISSUE TITLE
[REQ] Migrate off of com.google.code.findbugs:jsr305 or make it configurable

### ISSUE THREAD

**Issue by thokuest** [OpenAPITools/openapi-generator#7177]
### Is your feature request related to a problem? Please describe.

Pull request #3409 added nullable annotations to generated Java POJOs. The annotations are provided by library `com.google.code.findbugs:jsr305`. I'd like to cite [SpotBugs issue #130](https://github.com/spotbugs/spotbugs/issues/130) for why I think this is a problem for applications:

> JSR-305 is dormant, has been for a long while and shows no hope of ever producing an agreed set of annotations in our lifetime. Further more these annotations use `javax.` packages which it is not possible to use according to the [Oracle Java binary licence](http://www.oracle.com/technetwork/java/javase/terms/license/index.html), so applications can not use and ship these dependencies along with a JRE without violating the Oracle licence agreement.
> 
> > F. JAVA TECHNOLOGY RESTRICTIONS. You may not create, modify, or change the behavior of, or authorize your licensees to create, modify, or change the behavior of, classes, interfaces, or subpackages that are in any way identified as "java", "javax", "sun", “oracle” or similar convention as specified by Oracle in any naming convention designation.
> 
> The JSR-305 group has not defined any official releases according to its [jsr page](https://jcp.org/en/jsr/detail?id=305) so the only implementations is a seemingly random implementation provided by the FindBugs team. Even if the team where experts on the JSR (which some where) they are not official as there has been no vote and are not available from the JSR hompage - so the `javax` package name restriction still applies.

Additionally, using the library with the Java Platform Module System results in a split package and the application [won't compile](https://blog.codefx.org/java/jsr-305-java-9/).

## Describe the solution you'd like

Make it configurable so that annotations can be enabled/disabled explicitly. Also consider moving to a friendlier alternative. The [Kotlin Java Interop Documentation](https://kotlinlang.org/docs/reference/java-interop.html#nullability-annotations) enumerates some options.

## Describe alternatives you've considered

/

## Additional context

* [SpotBugs issue #130](https://github.com/spotbugs/spotbugs/issues/130) 
* [Guava issue #2960](https://github.com/google/guava/issues/2960)

**Comment by SarvagyaGupta** [comment]
@thokuest Is there any update on this issue?

**Comment by wing328** [comment]
One way is to use customized templates (e.g. via the `-t` CLI option). Please give it a try.

**Comment by SarvagyaGupta** [comment]
Hi @wing328 , I am pretty new to this. So if I want to remove the @Nullable annotation, how would I do so?

**Comment by CMoH** [comment]
I've also run into this problem. I want to remove the annotation from the generated code, or provide a custom annotation instead of `javax.annotation.Nullable`

**Comment by Cs4r** [comment]
> I've also run into this problem. I want to remove the annotation from the generated code, or provide a custom annotation instead of `javax.annotation.Nullable`

@CMoH did you find a fix for this?

**Comment by CMoH** [comment]
> > I've also run into this problem. I want to remove the annotation from the generated code, or provide a custom annotation instead of `javax.annotation.Nullable`
> 
> @CMoH did you find a fix for this?

I had a problem running the generated code within OSGi, so I used the following bundle as maven dependency:
```
            <dependency>
                <groupId>org.apache.servicemix.bundles</groupId>
                <artifactId>org.apache.servicemix.bundles.jsr305</artifactId>
                <version>3.0.2_1</version>
            </dependency>
```

**Comment by Cs4r** [comment]
t

> > > I've also run into this problem. I want to remove the annotation from the generated code, or provide a custom annotation instead of `javax.annotation.Nullable`
> > 
> > 
> > @CMoH did you find a fix for this?
> 
> I had a problem running the generated code within OSGi, so I used the following bundle as maven dependency:
> 
> ```
>             <dependency>
>                 <groupId>org.apache.servicemix.bundles</groupId>
>                 <artifactId>org.apache.servicemix.bundles.jsr305</artifactId>
>                 <version>3.0.2_1</version>
>             </dependency>
> ```

that isn't a solution, it only references `com.google.code.findbugs:jsr305:2.0.0` [you can see it here](https://svn.apache.org/viewvc/servicemix/smx4/bundles/tags/org.apache.servicemix.bundles.jsr305-2.0.0_1/pom.xml?view=markup)

**Comment by CMoH** [comment]
> t
> 
> > > > I've also run into this problem. I want to remove the annotation from the generated code, or provide a custom annotation instead of `javax.annotation.Nullable`
> > > 
> > > 
> > > @CMoH did you find a fix for this?
> > 
> > 
> > I had a problem running the generated code within OSGi, so I used the following bundle as maven dependency:
> > ```
> >             <dependency>
> >                 <groupId>org.apache.servicemix.bundles</groupId>
> >                 <artifactId>org.apache.servicemix.bundles.jsr305</artifactId>
> >                 <version>3.0.2_1</version>
> >             </dependency>
> > ```
> 
> that isn't a solution, it only references `com.google.code.findbugs:jsr305:2.0.0` [you can see it here](https://svn.apache.org/viewvc/servicemix/smx4/bundles/tags/org.apache.servicemix.bundles.jsr305-2.0.0_1/pom.xml?view=markup)

You are correct. I ran into this problem when attempting to use the code in OSGi, so I applied this bundle switch as a workaround to the derived problem. I did not attempt to fix the root problem, i.e. the annoying annotation.

In these situations, I usually clone the offending project, develop my patch and submit it to the project devs. Since my own work does not afford to wait, I deploy a patched build to a local maven repo, perhaps using a different group ID, and use that while they debate the patch. Maybe you will find this idea helpful.

**Comment by Cs4r** [comment]
@CMoH, in order to fix the issue what I have done is the following:
- Use `exec-maven-plugin` to run a Python script after `mvn package`.
-  The Python script removes non standard `javax.annotation.*` from the generated classes.

#### Actual code:

**Mvn plugin config**
```xml
<plugin>
    <groupId>org.codehaus.mojo</groupId>
    <artifactId>exec-maven-plugin</artifactId>
    <version>3.2.0</version>
    <executions>
        <execution>
            <id>post-process</id>
            <phase>process-sources</phase>
            <goals>
                <goal>exec</goal>
            </goals>
            <configuration>
                <executable>python3</executable>
                <arguments>
                        <argument>${project.basedir}/post_process_actions.py</argument>
                </arguments>
            </configuration>
        </execution>
    </executions>
</plugin>
  ```
  
  **Python script**
  ```python
import os
import re

def remove_non_standard_javax_annotations():
    directory = "target/generated-sources/openapi/src/main/java/[YOUR_PACKAGE]"
    for root, dirs, files in os.walk(directory):
        for file in files:
            if file.endswith(".java"):
                filepath = os.path.join(root, file)
                with open(filepath, "r") as f:
                    content = f.read()
                # Remove @javax.annotation.* annotations (non-standard)
                content = re.sub(r'(@javax\.annotation\..*?)\n', '', content)
                content = re.sub(r'@javax\.annotation\..*?\s+', '', content)
                with open(filepath, "w") as f:
                    f.write(content)
  
  ```

**Comment by CMoH** [comment]
We also had some post-processing tasks to cleanup unwanted files from generated sources. However, we relied on `maven-antrun-plugin`, which keeps the build confined to the Java world:

```
            <plugin>
                <artifactId>maven-antrun-plugin</artifactId>
                <groupId>org.apache.maven.plugins</groupId>
                <version>1.7</version>
                <executions>
                    <execution>
                        <id>remove-swagger-codegen-files</id>
                        <phase>generate-sources</phase>
                        <goals>
                            <goal>run</goal>
                        </goals>
                        <configuration>
                            <target>
                                <delete file="${project.build.directory}/generated-sources/openapi/pom.xml"/>
                                <delete file="${project.build.directory}/generated-sources/openapi/README.md"/>
                                <delete file="${project.build.directory}/generated-sources/openapi/build.gradle"/>
                                <delete file="${project.build.directory}/generated-sources/openapi/settings.gradle"/>
                                <delete file="${project.build.directory}/generated-sources/openapi/gradle.properties"/>
                                <delete dir="${project.build.directory}/generated-sources/openapi/gradle"/>
                                <delete file="${project.build.directory}/generated-sources/openapi/gradlew"/>
                                <delete file="${project.build.directory}/generated-sources/openapi/gradlew.bat"/>
                                <delete file="${project.build.directory}/generated-sources/openapi/.gitignore"/>
                                <delete file="${project.build.directory}/generated-sources/openapi/git_push.sh"/>
                                <delete file="${project.build.directory}/generated-sources/openapi/.travis.yml"/>
                                <delete dir="${project.build.directory}/generated-sources/openapi/src"/>
                                <delete dir="${project.build.directory}/generated-sources/openapi/docs"/>
                                <delete file="${project.build.directory}/generated-sources/openapi/.swagger-codegen"/>
                                <delete file="${project.build.directory}/generated-sources/openapi/build.sbt"/>
                                <delete file="${project.build.directory}/generated-sources/openapi/annotations"/>
                                <delete dir="${project.build.directory}/generated-sources/openapi/.openapi-generator"/>
                                <delete file="${project.build.directory}/generated-sources/openapi/.openapi-generator-ignore"/>
                            </target>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
```

If reliance on python or other external tools needs to be avoided, I found some maven plugins that can replace text in files. Also, writing a maven plugin is not hard.

**Comment by T3rm1** [comment]
@wing328 I think it would be a good solution to either remove the findbugs dependency (and its annoations) all together or at least make it opt-in. We shouldn't have to work around this issue by either post processing or using custom templates. These annotations were questionable the whole time and by now it's been stated by various people that it's not recommend to use them any more.

I hope we can move forward here.

**Comment by T3rm1** [comment]
@wing328 Reminder

**Comment by wing328** [comment]
cc Java technical committee to see if they've an opinion on this:

@bbdouglas (2017/07) @sreeshas (2017/08) @jfiala (2017/08) @lukoyanov (2017/09) @cbornet (2017/09) @jeff9finger (2018/01) @karismann (2019/03) @Zomzog (2019/04) @lwlee2608 (2019/10) @martin-mfg (2023/08)

**Comment by martin-mfg** [comment]
#### TL;DR
I propose switching to [SpotBugs' `@Nullable` annotation](https://www.javadoc.io/doc/com.github.spotbugs/spotbugs-annotations/latest/edu/umd/cs/findbugs/annotations/Nullable.html). If this doesn't match your needs, please let us know.

#### Why?
Switching to annotations from a different dependency solves the problems cited by @thokuest, while still allowing us to generate annotated code. I am not sure which of the various alternatives to `com.google.code.findbugs.jsr305` we should use - I proposed spotbugs simply because I think it is well established. If anyone has good reasons to suggest a different alternative, please let us know.
Since switching to an alternative dependency should solve all cited problems, I don't see a reason to disable the generation of `@Nullable` annotations. But again, if there is a reason, please let us know. In this case it might make sense to make both changes - switch to spotbugs and also optionally skip the annotation.
And finally, if people agree to go with the spotbugs alternative, it would be great if someone could create a PR to implement the change. :)

**Comment by T3rm1** [comment]
> I don't see a reason to disable the generation of @Nullable annotations. But again, if there is a reason, please let us know.

Just for the sake of customizability. Let people decide for themselves which dependencies are added to their projects by the generated client. Don't force it. Just like Bean Validation is optional, too.

**Comment by Elewyth** [comment]
> #### TL;DR
> 
> I propose switching to [SpotBugs' `@Nullable` annotation](https://www.javadoc.io/doc/com.github.spotbugs/spotbugs-annotations/latest/edu/umd/cs/findbugs/annotations/Nullable.html). If this doesn't match your needs, please let us know.

I would suggest moving to [JSpecify](https://jspecify.dev/) instead. It is a joint initiative by [major players](https://jspecify.dev/), such as Google, Meta, Microsoft, Oracle, Sonar, to name a few. They aim to establish JSpecify as the new standard with clear specifications for null-analysis and nullability annotations. Version 1.0.0 was released in July 2024.

**Comment by T3rm1** [comment]
> > #### TL;DR
> > I propose switching to [SpotBugs' `@Nullable` annotation](https://www.javadoc.io/doc/com.github.spotbugs/spotbugs-annotations/latest/edu/umd/cs/findbugs/annotations/Nullable.html). If this doesn't match your needs, please let us know.
> 
> I would suggest moving to [JSpecify](https://jspecify.dev/) instead. It is a joint initiative by [major players](https://jspecify.dev/about/#who-are-we), such as Google, Meta, Microsoft, Oracle, Sonar, to name a few. They aim to establish JSpecify as the new standard with clear specifications for null-analysis and nullability annotations. Version 1.0.0 was released in July 2024.

I'd say this could be a temprary solution. It's more promising than SpotBugs but ultimately the JDK will include their own nullness markers. The topic can be tracked here: https://bugs.openjdk.org/browse/JDK-8303099 

So first step should be to get rid of FindBugs and add **optional** support for JSpeficy annotations. Later on, add support for Java's own nullness markers.

**Comment by martin-mfg** [comment]
Alright, apparently the consensus is JSpecify. Thanks for your input!

It would be great if someone could create a PR to switch from SpotBugs to JSpecify. And it would be great if someone could open a PR to make the JSpecify/SpotBugs support optional.

**Comment by Saljack** [comment]
I think that the best description to JSpecify is this well known xkcd stripe https://xkcd.com/927/ .

There is also `jakarta.annotation.Nullable` and `jakarta.annotation.Nonnull`. They are part of Jakarta Annotations so most of projects already have them so there is no additional dependency in  most of generator use cases.

**Comment by T3rm1** [comment]
> so most of projects already have them

You are wrong. Spring 7 uses JSpecify.
https://docs.spring.io/spring-framework/reference/7.0-SNAPSHOT/core/null-safety.html
